torch/nn/modules/activation.py

```diff
@@ -1258,20 +1258,27 @@ def merge_masks(self, attn_mask: Optional[Tensor], key_padding_mask: Optional[Te
         mask_type: Optional[int] = None
         merged_mask: Optional[Tensor] = None
 
-        if attn_mask is not None:
-            mask_type = 0
-            merged_mask = attn_mask
         if key_padding_mask is not None:
             mask_type = 1
             merged_mask = key_padding_mask
-        if (attn_mask is not None) and (key_padding_mask is not None):
+
+        if attn_mask is not None:
             # In this branch query can't be a nested tensor, so it has a shape
             batch_size, seq_len, _ = query.shape
             mask_type = 2
-            key_padding_mask_expanded = key_padding_mask.view(batch_size, 1, 1, seq_len) \
-                                                        .expand(-1, self.num_heads, -1, -1)
-            attn_mask_expanded = attn_mask.view(1, 1, seq_len, seq_len).expand(batch_size, self.num_heads, -1, -1)
-            merged_mask = attn_mask_expanded + key_padding_mask_expanded
+
+            # Always expands attn_mask to 4D
```

Comment:
Are we sure this doesn't introduce any regressions further down the stack if there's assumptions around the mask being 2D?

Reply:
It's passed to `masked_softmax` downstream along with the `mask_type`, which indicates an expected shape / format. AFAICT as long as the mask type fits with the output, we're good.

Reply:
@jbschlosser - I don't think so. There's a fast path within the fast path. `masked_softmax` has another check here: https://github.com/pytorch/pytorch/blob/ae55619a2bb73ebcdc80b02a6ccd72275a9ce23e/aten/src/ATen/native/cuda/SoftMax.cu#L970

Reply:
we have inception fast paths? sigh.

so this PR may introduce a perf regression for 2D masks but no correctness issues, right?

Reply:
Yes, I think so.

Reply:
As discussed offline, I don't think there is a regression after all. The fast-path kernel linked above is for `mask_type ==1`, which is when only a `key_padding_mask` is set, and the fix in this PR doesn't touch that case.

Interestingly, there's a TODO to write a fast-path kernel for `mask_type == 0`, `attn_mask` only set: https://github.com/pytorch/pytorch/blob/ae55619a2bb73ebcdc80b02a6ccd72275a9ce23e/aten/src/ATen/native/cuda/SoftMax.cu#L974-L976

Reply:
Yep, you're right. My bad. Still happy I pointed this out since these conditions are complex.